### PR TITLE
Register destructor callback is now templated

### PR DESCRIPTION
**Pull request by bastiankoe**
Added a template based register destructor callback function to the memory_object class.

### CODE REVIEW

File: include/boost/compute/memory_object.hpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 #include <boost/compute/kernel.hpp>
 #include <boost/compute/detail/get_object_info.hpp>
 #include <boost/compute/detail/assert_cl_success.hpp>
+#include <boost/bind.hpp>
```
Comment (kylelutz) [inline include/boost/compute/memory_object.hpp]: Hmm, I don't think this `#include` is necessary (looks like it's just used by the test)?

**Comment by kylelutz** [comment]
Just one small issue with the extra include. Also, could you rebase this on `develop` as I just merged your other pull-request which also touched `test_buffer.cpp` and now there's a conflict. Thanks!

File: include/boost/compute/memory_object.hpp (hunk 1 above)
Comment (bastiankoe) [inline include/boost/compute/memory_object.hpp]: Yeah it's not necessary. I don't know why I put this there, you're right that this belongs to the test.

**Comment by bastiankoe** [comment]
I hope that now everything is okay ;)

**Comment by kylelutz** [comment]
Merged! Thanks!

**Pull request merged**